README.md
# vue-custom-element-rails

[vue\-custom\-element](https://github.com/karol-f/vue-custom-element) wrapper

## Installation

```
yarn add @sonicgarden/vue-custom-element-rails
```

```ruby
gem 'vue_custom_element_rails', github: 'SonicGarden/vue-custom-element-rails'
```

## Usage

```javascript
import registerCustomElement from '@sonicgarden/vue-custom-element-rails'

// When IE11 support is not required
// import registerCustomElement from '@sonicgarden/vue-custom-element-rails/dist/slim'

import MyComponent from './components/MyComponent.vue'

registerCustomElement('my-component', MyComponent)
```

```haml
%my-component{ vue_component_props(user: user_hash, items: items) }
```

### With vue-i18n

```javascript
import Vue from 'vue'
import VueI18n from 'vue-i18n'
import registerCustomElement from 'vue-custom-element-rails'
import MyComponent from './components/MyComponent.vue'

Vue.use(VueI18n)

const messages = {
  en: {
    message: {
      hello: 'hello world',
    },
  },
  ja: {
    message: {
      hello: 'こんにちは、世界',
    },
  },
}

const i18n = new VueI18n({
  locale: 'ja',
  messages,
})

registerCustomElement('my-component', MyComponent, { vueOptions: { i18n } })
```

### Slot

In app/views/home/index.html.haml:

```haml
%my-component
  %p hello
```

```javascript
import registerCustomElement from 'vue-custom-element-rails'
import MyComponent from './components/MyComponent.vue'

registerCustomElement('my-component', MyComponent, { hasSlot: true })
```


CHANGELOG.md
## 0.2.1

- Support nil value

## 0.1.0

- First release


spec/spec_helper.rb
require "bundler/setup"
require "vue_custom_element_rails"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/vue/custom/element/rails_spec.rb
RSpec.describe Vue::Custom::Element::Rails do
  it "has a version number" do
    expect(Vue::Custom::Element::Rails::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


spec/vue_custom_element_rails/helpers_spec.rb
require 'active_support/core_ext/hash/keys' # NOTE: dekiru側で読み込むべき
require 'vue_custom_element_rails/helpers'

RSpec.describe VueCustomElementRails::Helpers do
  let(:helper) do
    (Class.new do
      include VueCustomElementRails::Helpers
    end).new
  end

  describe '#vue_component_props' do
    subject { helper.vue_component_props(props) }

    context 'underscore props' do
      let(:props) do
        {
          vue_custom_element: {
            user_name: 'hoge',
          },
          gem_version: 0.1,
        }
      end

      it { is_expected.to eq({ ':vue-custom-element' => '{"userName":"hoge"}', 'gem-version' => '0.1' }) }
    end

    context 'camelcase props' do
      let(:props) do
        {
          vueCustomElement: {
            user_name: 'hoge',
          },
          gem_version: 0.1,
        }
      end

      it { is_expected.to eq({ ':vue-custom-element' => '{"userName":"hoge"}', 'gem-version' => '0.1' }) }
    end

    context 'boolean props' do
      let(:props) do
        {
          is_show: true,
          is_hide: false,
        }
      end

      it { is_expected.to eq({ 'is-show' => 'true', 'is-hide' => 'false' }) }
    end

    context 'number props' do
      let(:props) do
        {
          integer: 10,
          float: 0.01,
        }
      end

      it { is_expected.to eq({ 'integer' => '10', 'float' => '0.01' }) }
    end

    context 'nil props' do
      let(:props) do
        {
          nil_value: nil,
          some_value: 'some'
        }
      end

      it { is_expected.to eq({ 'some-value' => 'some' }) }
    end
  end
end


src/slim.js
import Vue from 'vue'
import vueCustomElement from 'vue-custom-element'
import { getProps } from 'vue-custom-element/src/utils/props'
import camelCase from 'lodash.camelcase'

Vue.use(vueCustomElement)

function connectedCallback() {
  const el = this
  // NOTE: Object, Array型のプロパティを自前でJSON.parseする必要がある
  Array.from(el.attributes)
    .filter(({ name }) => name.startsWith(':'))
    .forEach(({ name, value }) => {
      el[camelCase(name.slice(1))] = JSON.parse(value)
    })
}

// NOTE: vue-custom-element && Chromeで発生するslot不具合の暫定対応 (遅延させないとchidNodesが取れない)
// SEE: https://github.com/karol-f/vue-custom-element/issues/162
const asyncComponentDefinition = component => () => new Promise((resolve) => {
  setTimeout(() => {
    resolve(component)
  }, 0)
})

const registerCustomElement = (
  tag,
  componentDefinition,
  { hasSlot, vueOptions, ...options } = {},
) => {
  // NOTE: Vue.extendで定義しているコンポーネントをそのまま渡すとasyncコンポーネント扱いになってエラーとなる為extendsOptionsを渡すようにしている
  // SEE: https://github.com/karol-f/vue-custom-element/issues/143
  // SEE: https://github.com/vuejs/vue/blob/v2.6.10/src/core/global-api/extend.js#L19-L81
  const componentOptions = componentDefinition.extendOptions
    ? componentDefinition.extendOptions
    : componentDefinition

  // NOTE: slotがある場合は遅延コンポーネント扱いしている為、props定義を自前で取得して配列形式で渡す必要がある
  const propsOptions = hasSlot ? { props: getProps(componentOptions).camelCase } : {}

  const beforeCreateVueInstance = rootElement => ({
    ...rootElement,
    ...vueOptions,
  })

  Vue.customElement(tag, hasSlot ? asyncComponentDefinition(componentOptions) : componentOptions, {
    ...options,
    ...propsOptions,
    connectedCallback,
    beforeCreateVueInstance,
  })
}

export default registerCustomElement


lib/vue_custom_element_rails/helpers.rb
require 'active_support/core_ext/string'
require 'dekiru/camelize_hash'
require 'json'

module VueCustomElementRails
  module Helpers
    def vue_component_props(props)
      hash_arr =
        props.map do |key, value|
          if value.is_a?(Hash) || value.is_a?(Array)
            key = ":#{key}"
            value = _vue_component_props_json(value)
          end

          [key.to_s.underscore.downcase.dasherize, value.nil? ? nil : value.to_s]
        end
      hash_arr.to_h.compact
    end

    using Dekiru::CamelizeHash

    def _vue_component_props_json(value)
      case value
      when Hash
        value.deep_camelize_keys(:lower).to_json
      when Array
        value.map { |v| v.is_a?(Hash) ? v.deep_camelize_keys(:lower) : v }.to_json
      else
        value
      end
    end
  end
end


src/slim.ts
import Vue, { ComponentOptions, VueConstructor } from 'vue'
import vueCustomElement from 'vue-custom-element'
import { getProps } from 'vue-custom-element/src/utils/props'
import camelCase from 'lodash.camelcase'

interface Options extends vueCustomElement.options {
  hasSlot?: boolean;
  vueOptions?: ComponentOptions<Vue>;
}

interface IregisterCustomElement {
  (tag: string, componentDefinition: ComponentOptions<Vue>, options?: Options): void;
  (tag: string, singleFileComponent: VueConstructor<Vue>, options?: Options): void;
  (tag: string, asyncComponentDefinition: () => Promise<ComponentOptions<Vue>>, options?: Options): void;
}

Vue.use(vueCustomElement)

function connectedCallback() {
  const el = this as HTMLElement
  // NOTE: Object, Array型のプロパティを自前でJSON.parseする必要がある
  Array.from(el.attributes)
    .filter(({ name }) => name.startsWith(':'))
    .forEach(({ name, value }) => {
      el[camelCase(name.slice(1))] = JSON.parse(value)
    })
}

// NOTE: vue-custom-element && Chromeで発生するslot不具合の暫定対応 (遅延させないとchidNodesが取れない)
// SEE: https://github.com/karol-f/vue-custom-element/issues/162
const asyncComponentDefinition = (component) => () => new Promise((resolve) => {
  setTimeout(() => {
    resolve(component)
  }, 0)
})

const registerCustomElement: IregisterCustomElement = (
  tag: string,
  componentDefinition,
  { hasSlot, vueOptions, ...options } = { hasSlot: false, vueOptions: {} },
) => {
  // NOTE: Vue.extendで定義しているコンポーネントをそのまま渡すとasyncコンポーネント扱いになってエラーとなる為extendsOptionsを渡すようにしている
  // SEE: https://github.com/karol-f/vue-custom-element/issues/143
  // SEE: https://github.com/vuejs/vue/blob/v2.6.10/src/core/global-api/extend.js#L19-L81
  const componentOptions = componentDefinition.extendOptions
    ? componentDefinition.extendOptions
    : componentDefinition

  // NOTE: slotがある場合は遅延コンポーネント扱いしている為、props定義を自前で取得して配列形式で渡す必要がある
  const propsOptions: { props?: ComponentOptions<Vue>['props'] } = hasSlot
    ? { props: getProps(componentOptions).camelCase }
    : {}

  const beforeCreateVueInstance = (
    rootElement: ComponentOptions<Vue>,
  ): ComponentOptions<Vue> => ({
    ...rootElement,
    ...vueOptions,
  })

  Vue.customElement(
    tag,
    hasSlot ? asyncComponentDefinition(componentOptions) : componentOptions,
    {
      ...options,
      ...propsOptions,
      connectedCallback,
      beforeCreateVueInstance,
    },
  )
}

export default registerCustomElement
